### PR TITLE
Updates and tests for plan serialization and deserialization

### PR DESCRIPTION
**Pull request by lemillermicrosoft**
### Motivation and Context
The change improves the serialization and deserialization of Plan objects.

### Description
Allows deserialization of Plans without a context. Adds unit tests for the Plan and ContextVariables classes, which cover various scenarios and edge cases of serializing and deserializing plans and context variables.

Details:
- Updates the custom converter for the ContextVariables class, which serializes and deserializes the key-value pairs as an array of objects, instead of a dictionary. This improves the efficiency and readability of the JSON format.
- Add a new file PlanSerializationTests.cs to the SemanticKernel.UnitTests project, which contains unit tests for the Plan class. The tests verify that a plan can be serialized and deserialized correctly, and that the plan state is updated correctly after each step execution. The tests also check for invalid JSON strings.
- Add a new file ContextVariablesConverterTests.cs to the SemanticKernel.UnitTests project, which contains unit tests for the ContextVariablesConverter class. The tests verify that the converter can correctly serialize and deserialize context variables to and from JSON, and handle null, empty, and invalid inputs.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.UnitTests/Planning/SKContextExtensionsTests.cs [hunk 1]
```diff
@@ -42,7 +48,7 @@ public async Task CanCallGetAvailableFunctionsWithNoFunctionsAsync()
         var semanticQuery = "test";
 
         // Act
-        var result = await context.GetAvailableFunctionsAsync(config, semanticQuery);
+        var result = await context.GetAvailableFunctionsAsync(config, semanticQuery).ConfigureAwait(false);
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.UnitTests/Planning/SKContextExtensionsTests.cs:51]: No need for the ConfigureAwait in the test.

Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel.UnitTests/Planning/SKContextExtensionsTests.cs:51]: My IDE automatically does this so not going to worry about it right now.

File: dotnet/src/SemanticKernel/Orchestration/Plan.cs [hunk 2]
```diff
@@ -22,6 +22,12 @@ namespace Microsoft.SemanticKernel.Orchestration;
 /// </summary>
 public sealed class Plan : ISKFunction
 {
+    /// <summary>
+    /// The Schema version of the plan.
+    /// </summary>
+    [JsonPropertyName("schema")]
+    public int Schema { get; set; } = CurrentSchemaVersion;
```
Comment (dluc) [inline dotnet/src/SemanticKernel/Orchestration/Plan.cs:29]: what's the design about schemas? how are we detecting breaking changes, documenting them, handling upgrades, etc? do we have tests to ensure schemas are not broken?

**Review by dluc — CHANGES_REQUESTED** [review]
We have very limited knowledge about this schema, if it works for basic planner, and we haven't tested it with other scenarios, e.g. the tagging method that @umeshma has been using.

If we introduce versioning we need:
* test it with other scenarios and agree that the schema is stable and will not change, which includes reviewing and documenting it
* testing each version to verify the contract remains intact from release to release
* decide how to upgrade from version to version when the time comes to have a new version
* documenting versions, breaking changes, SLA. If this is V1, are we going to maintain it for how long?
* how should users handle versions? why are there versions to worry about in the first place?

File: dotnet/src/SemanticKernel/Orchestration/Plan.cs (hunk 2 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Orchestration/Plan.cs:29]: There's more details to figure out, this is a stake in the ground similar to PromptTemplateConfig to make sure we can easily identify the serialization format of the plan and handle any changes in the future.

The schema version is used to validate the JSON serialization and deserialization of the plan, and an exception is thrown if the schema version does not match the current supported version. 

At the time of authoring, it wasn't a breaking change since this is all new. That's changed, but I think the impact is negligible right now, still. In the future, these things will be important, agreed.

The PR adds unit tests for both the Plan and ContextVariables classes, which cover various scenarios and edge cases of serializing and deserializing plans and context variables, including checks for invalid schema versions and invalid JSON strings.

Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Orchestration/Plan.cs:29]: Will comment out schema work for now and leave for future/finalization to leave this PR to focus on improvements.

File: dotnet/src/SemanticKernel.UnitTests/Planning/SKContextExtensionsTests.cs (hunk 1 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel.UnitTests/Planning/SKContextExtensionsTests.cs:51]: Actually I'll fix this too.

File: dotnet/src/SemanticKernel/Orchestration/Plan.cs (hunk 2 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Orchestration/Plan.cs]: Remove schema changes for now.

File: dotnet/src/SemanticKernel.UnitTests/Planning/SKContextExtensionsTests.cs (hunk 1 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel.UnitTests/Planning/SKContextExtensionsTests.cs]: #655

**Review by shawncal — APPROVED** [review]

**Review by dluc — CHANGES_REQUESTED** [review]
could you hold on until we merge the PR touching namespaces?

**Comment by lemillermicrosoft** [comment]
Currently holding for #598

**Review by dluc — APPROVED** [review]
